src/NServiceBus.Core/Recoverability/SecondLevelRetries/SecondLevelRetriesBehavior.cs

```diff
@@ -30,39 +43,50 @@ public SecondLevelRetriesBehavior(SecondLevelRetryPolicy retryPolicy, string loc
+                failureInfoStorage.MarkForDeferralForSecondLevelRetry(context.Message.MessageId, ExceptionDispatchInfo.Capture(ex));
 
-                TimeSpan delay;
+                context.AbortReceiveOperation();
+            }
+        }
 
-                if (retryPolicy.TryGetDelay(message, ex, currentRetry, out delay))
-                {
-                    message.RevertToOriginalBodyIfNeeded();
-                    var messageToRetry = new OutgoingMessage(message.MessageId, message.Headers, message.Body);
+        async Task DeferMessageForSecondLevelRetry(ITransportReceiveContext context, Func<IRoutingContext, Task> fork, IncomingMessage message, ProcessingFailureInfo failureInfo)
+        {
+            var currentRetry = GetNumberOfRetries(message.Headers) + 1;
 
-                    messageToRetry.Headers[Headers.Retries] = currentRetry.ToString();
-                    messageToRetry.Headers[RetriesTimestamp] = DateTimeExtensions.ToWireFormattedString(DateTime.UtcNow);
+            message.Headers[Headers.FLRetries] = (failureInfo.FailedProcessingAttempts - 1).ToString();
 
-                    var dispatchContext = this.CreateRoutingContext(messageToRetry, localAddress, context);
+            TimeSpan delay;
 
-                    context.Extensions.Set(new List<DeliveryConstraint>
-                    {
-                        new DelayDeliveryWith(delay)
-                    });
+            if (retryPolicy.TryGetDelay(message, failureInfo.Exception, currentRetry, out delay))
+            {
+                message.RevertToOriginalBodyIfNeeded();
+                var messageToRetry = new OutgoingMessage(message.MessageId, message.Headers, message.Body);
 
-                    Logger.Warn($"Second Level Retry will reschedule message '{message.MessageId}' after a delay of {delay} because of an exception:", ex);
+                messageToRetry.Headers[Headers.Retries] = currentRetry.ToString();
+                messageToRetry.Headers[Headers.RetriesTimestamp] = DateTimeExtensions.ToWireFormattedString(DateTime.UtcNow);
 
-                    await fork(dispatchContext).ConfigureAwait(false);
+                var dispatchContext = this.CreateRoutingContext(messageToRetry, localAddress, context);
 
-                    await context.RaiseNotification(new MessageToBeRetried(currentRetry, delay, context.Message, ex)).ConfigureAwait(false);
+                context.Extensions.Set(new List<DeliveryConstraint>
+                {
+                    new DelayDeliveryWith(delay)
+                });
 
-                    return;
-                }
+                Logger.Warn($"Second Level Retry will reschedule message '{message.MessageId}' after a delay of {delay} because of an exception:", failureInfo.Exception);
+
+                await fork(dispatchContext).ConfigureAwait(false);
+
+                failureInfoStorage.ClearFailureInfoForMessage(message.MessageId);
```

Comment:
@tmasternak commiting the transaction can fail after `ClearFailureInfoForMessage` has been called. This will result in the message being redelivered but the failure storage being empty for that message, meaning it will be retried again starting with FLR. Seems to be a tradeoff on different levels (performance, memory pressure, "number of retries consistency").